tensorflow/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -1044,12 +1045,29 @@ class CudnnRnnDescriptor : public dnn::RnnDescriptor {
     cudnnRNNAlgo_t rnn_algo = ToCudnnRNNAlgo(algorithm_config.algorithm());
 
     // TODO: allow the user to choose an algorithm.
-    RETURN_IF_CUDNN_ERROR(cudnnSetRNNDescriptor_v6(
-        cudnn.handle(), /*rnnDesc=*/rnn_desc.get(), /*hiddenSize=*/hidden_size,
-        /*numLayers=*/num_layers, /*dropoutDesc=*/dropout_desc.handle(),
-        /*inputMode=*/input_mode, /*direction=*/direction_mode,
-        /*mode=*/rnn_mode, /*algo=*/rnn_algo,
-        /*dataType=*/compute_type));
+    if (c_size != 0 && hidden_size < c_size) {
+      RETURN_IF_CUDNN_ERROR(cudnnSetRNNDescriptor_v6(
+          cudnn.handle(), /*rnnDesc=*/rnn_desc.get(), /*hiddenSize=*/c_size,
+          /*numLayers=*/num_layers, /*dropoutDesc=*/dropout_desc.handle(),
+          /*inputMode=*/input_mode, /*direction=*/direction_mode,
+          /*mode=*/rnn_mode, /*algo=*/rnn_algo, /*dataType=*/compute_type));
+#if CUDNN_VERSION >= 7101
+      RETURN_IF_CUDNN_ERROR(cudnnSetRNNProjectionLayers(
+          cudnn.handle(), /*rnnDesc=*/rnn_desc.get(),
+          /*recProjSize=*/hidden_size, /*outProjSize=*/0));
+#else
+      return port::Status(port::error::INVALID_ARGUMENT,
+                          "No supported cudnnSetRNNProjectionLayers when "
+                          "CUDNN_VERSION < 7.1.1");
+#endif
+    } else {
+      RETURN_IF_CUDNN_ERROR(cudnnSetRNNDescriptor_v6(
```

Comment:
Please avoid code duplication. We can use a local variable to record the actual `hiddenSize` before calling.